mobile_verifier/src/speedtests.rs

```diff
@@ -99,577 +85,107 @@ impl SpeedtestDaemon {
-        );
-        assert_eq!(
-            Average::from(&speedtests[4..=5]).tier(),
-            SpeedtestTier::Acceptable
-        );
-        assert_eq!(
-            Average::from(&speedtests[4..=6]).tier(),
-            SpeedtestTier::Acceptable
-        );
-    }
-
-    #[test]
-    fn check_minimum_known_invalid() {
-        let speedtests = known_speedtests();
-        assert_ne!(
-            Average::from(&speedtests[5..6]).tier(),
-            SpeedtestTier::Acceptable
-        );
-    }
-
-    #[test]
-    fn check_speedtest_rolling_avg() {
-        let owner: PublicKeyBinary = "112NqN2WWMwtK29PMzRby62fDydBJfsCLkCAf392stdok48ovNT6"
-            .parse()
-            .expect("failed owner parse");
-        let speedtests = VecDeque::from(known_speedtests());
-        let avgs = SpeedtestAverages {
-            speedtests: HashMap::from([(owner, speedtests)]),
-        }
-        .into_iter();
-        for avg in avgs {
-            if let Some(first) = avg.speedtests.first() {
-                assert_eq!(avg.latest_timestamp, first.timestamp);
-            }
-        }
-    }
-
-    #[test]
-    fn check_speedtest_without_lapsed() {
-        let speedtest_cutoff = Duration::hours(10);
-        let contiguos_speedtests = known_speedtests();
-        let contiguous_speedtests =
-            speedtests_without_lapsed(contiguos_speedtests.iter(), speedtest_cutoff);
-
-        let disjoint_speedtests = vec![
-            Speedtest::new(
-                parse_dt("2022-08-02 6:00:00 +0000"),
-                bytes_per_s(20),
-                bytes_per_s(150),
-                70,
-            ),
-            Speedtest::new(
-                parse_dt("2022-08-01 18:00:00 +0000"),
-                bytes_per_s(10),
-                bytes_per_s(118),
-                50,
-            ),
-            Speedtest::new(
-                parse_dt("2022-08-01 12:00:00 +0000"),
-                bytes_per_s(30),
-                bytes_per_s(112),
-                40,
-            ),
-        ];
-        let disjoint_speedtests =
-            speedtests_without_lapsed(disjoint_speedtests.iter(), speedtest_cutoff);
-
-        assert_eq!(contiguous_speedtests.count(), 8);
-        assert_eq!(disjoint_speedtests.count(), 1);
-    }
+pub async fn save_speedtest(
+    speedtest: &CellSpeedtest,
+    exec: &mut Transaction<'_, Postgres>,
+) -> Result<(), sqlx::Error> {
+    sqlx::query(
+        r#"
+        insert into speedtests (pubkey, upload_speed, download_speed, latency, serial, timestamp)
+        values ($1, $2, $3, $4, $5, $6)
+        on conflict (pubkey, timestamp) do nothing
+        "#,
+    )
+    .bind(speedtest.pubkey.clone())
+    .bind(speedtest.upload_speed as i64)
+    .bind(speedtest.download_speed as i64)
+    .bind(speedtest.latency as i64)
+    .bind(speedtest.serial.clone())
+    .bind(speedtest.timestamp)
+    .execute(exec)
+    .await?;
+    Ok(())
+}
+
+pub async fn get_latest_speedtests_for_pubkey(
```

Comment:
Since you’re not performing any operations on the steamed result in the `while` loop you should be able to just make this a `fetch_all` instead of a `fetch` and get the results back in a vec without the while loop

Reply:
addressed here: 67effc3558d2cd2c56ef07956dba0b19b02ffd95